File: backend/src/main/java/com/rbcits/backend/services/CategoryService.java
package com.rbcits.backend.services;

import com.rbcits.backend.DTOs.CategoryDTO;
import com.rbcits.backend.models.Category;
import com.rbcits.backend.repositories.CategoryRepository;
import org.springframework.stereotype.Service;
import java.util.List;
import java.util.stream.Collectors;

@Service
public class CategoryService {

    private final CategoryRepository categoryRepository;

    public CategoryService(CategoryRepository categoryRepository) {
        this.categoryRepository = categoryRepository;
    }

    public CategoryDTO createCategory(CategoryDTO dto) {
        Category category = new Category();
        category.setName(dto.getName());
        Category saved = categoryRepository.save(category);
        return convertToDTO(saved);
    }

    public CategoryDTO getCategoryById(Long id) {
        Category category = categoryRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Category not found"));
        return convertToDTO(category);
    }

    public List<CategoryDTO> getAllCategories() {
        return categoryRepository.findAll().stream()
                .map(this::convertToDTO)
                .collect(Collectors.toList());
    }

    public CategoryDTO updateCategory(Long id, CategoryDTO dto) {
        Category category = categoryRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Category not found"));
        category.setName(dto.getName());
        Category updated = categoryRepository.save(category);
        return convertToDTO(updated);
    }

    public void deleteCategory(Long id) {
        if (!categoryRepository.existsById(id)) {
            throw new IllegalArgumentException("Category not found");
        }
        categoryRepository.deleteById(id);
    }

    private CategoryDTO convertToDTO(Category category) {
        return new CategoryDTO(category.getCategory_id(), category.getName());
    }
}

File: frontend/src/services/categoryService.ts
import api from './api'

export interface Category {
  categoryId: number
  name: string
}

const categoryService = {
  createCategory: async (name: string): Promise<Category> => {
    const response = await api.post('/categories', { name })
    return response.data
  },

  getCategoryById: async (id: number): Promise<Category> => {
    const response = await api.get(`/categories/${id}`)
    return response.data
  },

  getAllCategories: async (): Promise<Category[]> => {
    const response = await api.get('/categories')
    return response.data
  },

  updateCategory: async (id: number, name: string): Promise<Category> => {
    const response = await api.put(`/categories/${id}`, { name })
    return response.data
  },

  deleteCategory: async (id: number): Promise<void> => {
    await api.delete(`/categories/${id}`)
  },
}

export default categoryService
